batch/sql/add-lock-in-share-mode.sql

```diff
@@ -0,0 +1,307 @@
+/*
+380c26
+<     SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name FOR UPDATE;
+---
+>     SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name LOCK IN SHARE MODE;
+422c68
+<   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name FOR UPDATE;
+---
+>   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name LOCK IN SHARE MODE;
+473c119
+<   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name FOR UPDATE;
+---
+>   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name LOCK IN SHARE MODE;
+525c171
+<   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name FOR UPDATE;
+---
+>   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name LOCK IN SHARE MODE;
+574c220
+<   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name FOR UPDATE;
+---
+>   SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name LOCK IN SHARE MODE;
+*/
+
+DELIMITER $$
+
+DROP PROCEDURE IF EXISTS add_attempt;
+CREATE PROCEDURE add_attempt(
+  IN in_batch_id BIGINT,
+  IN in_job_id INT,
+  IN in_attempt_id VARCHAR(40),
+  IN in_instance_name VARCHAR(100),
+  IN in_cores_mcpu INT,
+  OUT delta_cores_mcpu INT
+)
+BEGIN
+  DECLARE attempt_exists BOOLEAN;
+  DECLARE cur_instance_state VARCHAR(40);
+  SET delta_cores_mcpu = IFNULL(delta_cores_mcpu, 0);
+
+  SET attempt_exists = EXISTS (SELECT * FROM attempts
+                               WHERE batch_id = in_batch_id AND
+                                 job_id = in_job_id AND attempt_id = in_attempt_id
+                               FOR UPDATE);
+
+  IF NOT attempt_exists AND in_attempt_id IS NOT NULL THEN
+    INSERT INTO attempts (batch_id, job_id, attempt_id, instance_name) VALUES (in_batch_id, in_job_id, in_attempt_id, in_instance_name);
+    SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name LOCK IN SHARE MODE;
```

Comment:
This change seems wrong.  You update instances two lines below.  Why should this be in SHARE MODE?

Reply:
good catch!